std/stdio.d

```diff
@@ -1522,6 +1524,139 @@ for every line.
+            }
+
+            return buildPath(dir, name);
+        }
+
+        // Once should be enough given the randomness, but we'll give it three
+        // times to be safe.
+        foreach(i; 0 .. 3)
+        {
+            auto filename = genTempName(prefix, suffix, dir);
+
+            // O_EXCL does not support symbolic links.
+            if(dirName(filename).isSymlink)
+                filename = buildPath(absolutePath(dirName(filename), baseName(filename)));
+
+            import std.internal.cstring : tempCString;
+            version(Posix)
+            {
+                import core.sys.posix.sys.stat, core.sys.posix.fcntl;
+                enum flags = S_IRUSR | S_IWUSR;
+                auto fd = open(tempCString(filename), O_RDWR | O_CREAT | O_EXCL, flags);
+            }
+            else version(Windows)
+            {
+                enum flags = _S_IREAD | _S_IWRITE;
+                auto fd = _wsopen(tempCString!wchar(filename), O_RDWR | O_CREAT | O_EXCL, _SH_DENYNO, flags);
+            }
+            else static assert("Unsupported OS");
+
+            if(fd == -1)
+                continue;
+
+            File retval;
+            retval.fdopen(fd, "w+b", filename);
+            return retval;
+        }
+
+        throw new ErrnoException("Failed to create a temporary file.");
+    }
+
+    unittest
+    {
+        import std.algorithm, std.conv, std.file, std.path, std.process;
+
+        {
+            string name;
+            scope(exit) if(name.exists) std.file.remove(name);
+            {
+                auto file = File.tempFile();
+                name = file.name;
+                assert(!name.empty);
+                assert(buildNormalizedPath(name.dirName) == buildNormalizedPath(tempDir()));
+                assert(!name.baseName.empty);
+                assert(name.exists);
+                assert(name.isFile);
+                file.write("hello world");
+            }
+            assert(name.exists);
+            assert(name.isFile);
+            assert(readText(name) == "hello world");
+        }
+
+        auto dir = buildPath(tempDir(), "test_sub_dir_" ~ to!string(thisProcessID));
+        mkdir(dir);
+        scope(exit) if(dir.exists) rmdirRecurse(dir);
+        string name;
+        {
+            auto file = File.tempFile("foobar", ".baz", dir);
+            name = file.name;
+            assert(!name.empty);
+            assert(name.dirName == dir);
+            assert(buildNormalizedPath(name.dirName) == buildNormalizedPath(dir));
+            assert(!name.baseName.empty);
+            assert(name.baseName.startsWith("foobar"));
+            assert(name.extension == ".baz");
+            assert(name.length > "foobar".length + ".baz".length);
+            assert(name.exists);
+            assert(name.isFile);
+            file.write("D for the win!");
+        }
+        assert(name.exists);
+        assert(name.isFile);
+        assert(readText(name) == "D for the win!");
+    }
```

Comment:
There's a lot of asserts in a single unittest block. Could you split this up in multiple unittest blocks for different scenarios.

Perhaps a documented unittest block as well. This function hasn't any examples in the documentation.